PythonVisualizations/BinaryTreeBase.py

```diff
@@ -0,0 +1,515 @@
+      nodeFamily = set(self.canvas.find_withtag(node.tag))
+      # get all the lines that are left
+      lines = set(self.canvas.find_withtag("right line"))
+
+      return (lines & nodeFamily).pop()
+      
+   # returns a tuple of the left and right child of node
+   def getChildren(self, node):
+      return self.getLeftChild(node), self.getRightChild(node)
+
+   # returns a list of all the nodes that descend from node
+   def getAllDescendants(self, node):
+      if not node: return []
+
+      children = self.getChildren(node)
+      returnList = []
+      if children[0]: returnList.append(children[0])
+      if children[1]: returnList.append(children[1])
+
+      return returnList + self.getAllDescendants(children[0]) + self.getAllDescendants(children[1])
+      
+
+   # ----------- DRAWING METHODS -------------------
+   
+   # monkey patching to allow for circles to be drawn easier
+   def _create_circle(self, x, y, r, **kwargs):
+        return self.create_oval(x-r, y-r, x+r, y+r, **kwargs)
+   Canvas.create_circle = _create_circle
+
+   # Create an arrow to point at a node
+   def createArrow(self, node):
+      arrow = self.canvas.create_line(node.coords[0], node.coords[1] - self.CIRCLE_SIZE - self.ARROW_HEIGHT,
+                                   node.coords[0], node.coords[1] - self.CIRCLE_SIZE, arrow="last", fill='red')
+      return (arrow, )
+
+   # move the arrow to point above the node
+   def moveArrow(self, arrow, node, numSteps= 5):
+      # was a node passed in?
+      if isinstance(node, Node):
+            toPos = (node.coords[0], node.coords[1] - self.CIRCLE_SIZE - self.ARROW_HEIGHT,
+            node.coords[0], node.coords[1] - self.CIRCLE_SIZE)
+      # were the coords passed?
+      elif isinstance(node, tuple):
+            toPos = (node[0], node[1] - self.CIRCLE_SIZE - self.ARROW_HEIGHT,
+            node[0], node[1] - self.CIRCLE_SIZE)
+
+      self.moveItemsTo(arrow, (toPos,), steps = numSteps)
+
+   # calculate the coordinates for the node shape
+   def calculateCoordinates(self, parent, level, childDirection):
+        if level == 0:
+            return self.ROOT_X0, self.ROOT_Y0
+        elif childDirection == Child.LEFT:
+            return parent.coords[0] - 1/2**level* (self.NODE_X_GAP-self.CIRCLE_SIZE), self.ROOT_Y0+level* self.NODE_Y_GAP
+        else:
+            return parent.coords[0] + 1/2**level* (self.NODE_X_GAP-self.CIRCLE_SIZE), self.ROOT_Y0+level* self.NODE_Y_GAP
+
+   # calculate the coordinates for the line attached to the node
+   def calculateLineCoordinates(self, node):
+      x1 = node.coords[0]
+      y1 = node.coords[1] +self.CIRCLE_SIZE
+
+      # where would the child be drawn?
+      level = self.getLevel(node)+1
+      x2, y2 = self.calculateCoordinates(node, level, Child.LEFT)
+      x3, y3 = self.calculateCoordinates(node, level, Child.RIGHT)
+      y2 -= self.CIRCLE_SIZE
+      y3 -= self.CIRCLE_SIZE
+
+      return [(x1, y1, x2, y2), (x1, y1, x3, y3)]
```

Comment:
I think replacing all the references to `self.CIRCLE_SIZE` with 0 would make the lines terminate on the circle centers.

Reply:
Making the suggested change makes the lines visible in the center of the circle. Maybe we should fill the circles with a color?

Here is what it looks like with the change:
![image](https://user-images.githubusercontent.com/36013651/87602587-3ea3ca80-c6c5-11ea-9e00-ea81d6850893.png)

Reply:
Implemented in commit ce09b89
Colors to come in a separate commit

Reply:
Colors and circles on top of the lines are implemented in commit 020e9c6